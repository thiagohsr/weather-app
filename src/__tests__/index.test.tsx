import { render, screen } from "@testing-library/react";
import Home from "@pages/index";
import { renderWithProviders } from "./utils/test-store-utils";
import { rest } from "msw";
import { setupServer } from "msw/node";

import themingWrapper from "@helpers/themingHelper";

import { currentWeather } from "./fixtures/currentWeather.fixture";
import { queries } from "./fixtures/weatherApiQueries.fixture";
import { act } from "react-dom/test-utils";


// TODO: populate store from these calls.
export const handlers = [
  rest.get(`/weather`, (req, res, ctx) => {
    return res(ctx.json({ status: "pending" }), ctx.delay(350));
  }),
  rest.get(`/reverse`, (req, res, ctx) => {
    return res(
      ctx.json([]), ctx.delay(1500)
    );
  }),
];
const server = setupServer(...handlers);

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

// Disable API mocking after the tests are done.
afterAll(() => server.close());

describe("Home", () => {
  it("should renders homepage unchanged", () => {
    const { container } = renderWithProviders(themingWrapper(<Home />));

    expect(container.innerHTML).toMatchSnapshot();
  });

  it("should renders a heading", () => {
    const { container } = renderWithProviders(themingWrapper(<Home />));
    const heading = screen.getByRole("heading", {
      name: /Weather app/i,
    });

    expect(heading).toBeInTheDocument();
  });

  it("should renders weatherDisplay feature", async () => {
    const { container, store } = await renderWithProviders(themingWrapper(<Home />), {
      preloadedState: {
        weatherApi: {
          queries: {
            ...queries            
          },
        },
        weather: {
          weather: { ...currentWeather.weather.weather },  
          forecast: currentWeather.forecast,
        },
      },
    });

    await act(() => {
      const forecastElementsList = screen.getByTestId('forecast-list');

      Object.values(forecastElementsList.children).forEach((item, index) => {
        const expectedWeatherDescription = currentWeather.weather.weather.daily[index].weather[0].description;
        expect(item.innerHTML).toContain(expectedWeatherDescription);
      });
      const expectedCityName = screen.getByRole("heading", {
        name: /São Paulo, BR/i,
      });
      const expectedCityTemperature = screen.getByRole("heading", {
        name: /18°C/i,
      });
      const expectedForecastListHeader = screen.getByRole("heading", {
        name: /8-day forecast/i,
      });
      expect(expectedCityName).toBeInTheDocument();
      expect(expectedCityTemperature).toBeInTheDocument();
      expect(expectedForecastListHeader).toBeInTheDocument();
      
      
    });
  });
});
