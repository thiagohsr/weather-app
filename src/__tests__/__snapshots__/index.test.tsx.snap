// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders homepage unchanged 1`] = `"<div class="sc-1a67062-0 kaaXyl"><main class="sc-1a67062-1 gaQGMz"><h1 class="sc-1a67062-2 bmEpcN">Weather app</h1><a href="/citiesList">Favourited cities</a><div class="sc-6de51f50-0 jowzkp"><div style="color: rgb(17, 17, 17); height: 24px; width: 32px; position: absolute; right: 20px; top: 25px;"><svg width="24" height="24" viewBox="0 0 24 24" focusable="false" role="presentation"><path d="M16.436 15.085l3.94 4.01a1 1 0 0 1-1.425 1.402l-3.938-4.006a7.5 7.5 0 1 1 1.423-1.406zM10.5 16a5.5 5.5 0 1 0 0-11 5.5 5.5 0 0 0 0 11z" fill="currentColor" fill-rule="evenodd"></path></svg></div><input placeholder="Search some place..." class="sc-6de51f50-1 cxVRPx"><ul class="sc-6de51f50-2 lezpQt"></ul></div>Weather is Loading!</main></div>"`;
